src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -59,68 +83,52 @@ private bool GetNext()
+                    {
+                        IEnumerator<TSource> enumerator = enumerable.GetEnumerator();
+                        ++state;
+                        if (enumerator.MoveNext())
                         {
-                            _enumerator.Dispose();
-                            _enumerator = _second.GetEnumerator();
-                            if (!_enumerator.MoveNext())
-                            {
-                                Dispose();
-                                return false;
-                            }
-                            state = 3;
-                        }
-                        TSource element = _enumerator.Current;
-                        _set = new Set<TSource>(_comparer);
-                        _set.Add(element);
-                        current = element;
-                        return true;
-                    case 2:
-                        if (GetNext())
+                            SetEnumerator(enumerator);
+                            StoreFirst();
                             return true;
-                        _enumerator.Dispose();
-                        _enumerator = _second.GetEnumerator();
-                        state = 3;
-                        goto case 3;
-                    case 3:
+                        }
+                    }
+                }
+                else if (state > 0)
+                {
+                    for (;;)
+                    {
                         if (GetNext())
                             return true;
-                        break;
+                        IEnumerable<TSource> enumerable = GetEnumerable(state - 1);
+                        if (enumerable == null)
+                            break;
+                        SetEnumerator(enumerable.GetEnumerator());
+                        ++state;
+                    }
                 }
 
                 Dispose();
                 return false;
             }
 
-            public override void Dispose()
-            {
-                if (_enumerator != null)
-                {
-                    _enumerator.Dispose();
-                    _enumerator = null;
-                    _set = null;
-                }
-
-                base.Dispose();
-            }
-
             private Set<TSource> FillSet()
             {
                 Set<TSource> set = new Set<TSource>(_comparer);
-                foreach (TSource element in _first)
-                    set.Add(element);
-                foreach (TSource element in _second)
-                    set.Add(element);
+                IEnumerable<TSource> enumerable = GetEnumerable(0);
+                int index = 0;
+                do
+                {
+                    foreach (TSource item in enumerable)
+                    {
+                        set.Add(item);
+                    }
+                } while ((enumerable = GetEnumerable(++index)) != null);
```

Comment:
I went through that, `for(var enumerable = GetEnumerable(0); enumerable != null; enumerable = GetEnumerable(++index))` and the above. I decided on the above as sometimes people look sideways at any `for` that isn't `for(numeric_set_up; numeric_check; numeric_operation)` but I've no great love or hatred for any of them.